src/pow.rs

```diff
@@ -0,0 +1,1164 @@
+// Rust Bitcoin Library - Written by the rust-bitcoin developers.
+// SPDX-License-Identifier: CC0-1.0
+
+//! Proof-of-work related integer types.
+//!
+//! The functions here are designed to be fast.
+//!
+
+macro_rules! impl_from_and_impl {
+    ($ty:ident) => {
+        impl From<[u64; 4]> for $ty {
+            #[inline]
+            fn from(a: [u64; 4]) -> Self { $ty(Uint256(a)) }
+        }
+
+        impl From<u64> for $ty {
+            #[inline]
+            fn from(x: u64) -> Self { $ty(Uint256::from(x)) }
+        }
+
+        impl $ty {
+            /// TODO: Generate docs.
+            #[inline]
+            pub const fn from_array(a: [u64; 4]) -> $ty { $ty(Uint256(a)) }
+
+            /// TODO: Generate docs.
+            #[inline]
+            pub fn from_be_bytes(bytes: [u8; 32]) -> $ty { $ty(Uint256::from_be_bytes((bytes))) }
+
+            /// TODO: Generate docs.
+            #[inline]
+            pub fn from_be_slice(bytes: &[u8]) -> Result<$ty, ParseLengthError> {
+                let u = Uint256::from_be_slice(bytes)?;
+                Ok($ty(u))
+            }
+
+            /// TODO: Generate docs.
+            #[inline]
+            pub fn to_be_bytes(self) -> [u8; 32] { self.0.to_be_bytes() }
+        }
+    };
+}
+
+macro_rules! impl_core_traits {
+    ($ty:ident) => {
+        impl core::ops::Add<$ty> for $ty {
```

Comment:
`Target` shouldn't implement `Add` because adding targets is non-sensical for reason I explained in the other PR. Same for sub. `Mul<Self>` and `Div<Self>` and `Rem<Self>` as well but `Mul<{integer}>` and `Div<{integer}>` are probably fine.

For `Difficulty`, adding and subtracting seems fine, other math operations with `Self` aren't. Multiplication/division by scalars should be OK.

Reply:
Ah cool, thanks. I should have flagged this. I knew I didn't quite get which operations should be implemented for which types. Thanks.

Reply:
I'm stuck with this (from `block.rs`)
```rust
    /// Computes the popular "difficulty" measure for mining.
    pub fn difficulty(&self) -> u64 {
        (Target::max() / self.target()).low_u64()
    }

    /// Returns the total work of the block.
    pub fn work(&self) -> Difficulty {
        self.target().to_difficulty()
    }
```
I'd like to move that logic onto the `Target` type so we don't have to implement `Div<Self>` for `Target` but I've already got
```rust
    /// Converts this [`Target`] to a [`Difficulty`].
    pub fn to_difficulty(&self) -> Difficulty {
        // 2**256 / (target + 1) == ~target / (target+1) + 1    (eqn shamelessly stolen from bitcoind)
        let mut ret = !self.0;
        let mut ret1 = self.0;
        ret1.increment();
        ret = ret / ret1;
        ret.increment();
        Difficulty(ret)
    }
```
I'm confused by the words difficulty and work.

Reply:
I think the API of `Target` and `Difficulty` needs carefully checking by someone with more knowledge of Bitcoin than me. I'm really guessing with this PR.

Reply:
I always knew that roughly difficulty = 1/target but never the details. Got me thinking about it and I believe I figured it out:

If difficulty = 1/target then target 0 would imply infinite difficulty - impossible. But the comparison of the block hash against target uses `<=` operator which implies 0 target is finding preimage to sha256 which is not impossible, just maximally difficult. Therefore 1 is added making the range of results (0, 1].

We don't want to deal with floating point numbers in computers so we just multiply it by 2^256. Notably this overflows U256 when target is 0. The assumption is that we can never hit it because there will never be target 0. However I still think it may be a bug in the context of abstract thinking about Bitcoin.

Then there's the magic equation from Core:

* when numbers are represented in two's complement, you get the negative of a number by computing `!x + 1`
* therefore `!x = -x - 1`
* however we keep using unsigned integer, so we get multiplicative group of integers modulo `MAX + 1` arithmetic
* so !x == `(MAX + 1) + (-x - 1)`
* modifying right side of the equation: `!target / (target + 1) + 1 == (2^256 -target - 1) / (target + 1) + 1`
* `== (2^256 -target - 1) / (target + 1) + (target + 1)/(target + 1)`
* `==  (2^256 -target - 1 + target + 1) / (target + 1)`
* `== 2^256 /(target +1)` QED

However note that there's another overflow bug in division there `Target::MAX.to_difficulty()` would divide by 0.

What does it mean for the type system though? I think the unit of any intermediate step is meaningless and we should just internally work on raw `U256`. Therefore `to_difficulty` stays as you wrote above. The `difficulty` function looks wrong but maybe there's something that uses that representation? If yes we should probably have a `lossy_to_u64` method for that.

Regarding the overflow bugs, I guess division should be patched to return 0 on `Target::MAX` and `Difficulty` should contain `enum DifficultyInner { Value(U256), MaxPlusOne }`. And no, the overflow bugs are not a problem in Core because it only decodes compact target which can not represent `Target::Max` and has an explicit check for target 0 (returning 0).

Reply:
Thanks man, this all took me a fair bit of thinking today :)

> The difficulty function looks wrong but maybe there's something that uses that representation? If yes we should probably have a lossy_to_u64 method for that.

I think this is aswered by https://bitcoin.stackexchange.com/questions/8806/what-is-difficulty-and-how-it-relates-to-target

I've updated the PR quite a bit to reflect my current understanding, please pick holes in it if you see them.

Reply:
Currently we have:
```rust
    fn inverse(&self) -> Uint256 {
        // We should never have a target/work of zero so this doesn't matter that much but we define
        // the inverse of 0 as MAX.
        if *self == Uint256::ZERO {
            return Uint256::MAX;
        }

        // We define the inverse of max as 1.
        if *self == Uint256::MAX {
            return Uint256::ONE;
        }

        // And the inverse of 1 as max.
        if *self == Uint256::ONE {
            return Uint256::MAX;
        }

        let mut denom = *self;
        denom.increment();

        let mut ret = !*self / denom;
        ret.increment();

        ret
    }
```